designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,162 @@
+
+More generally, Karpenter will need to guard against the following cases during a Karpenter upgrade:
+
+1. A `NodePool/EC2NodeClass` adds default values to existing fields that are included in the hash calculations
+2. A `NodePool/EC2NodeClass` field is added to the hash calculation with an already-set value
+3. A `NodePool/EC2NodeClass` field is removed from the hash calculations
+
+## Consideration
+
+Karpenter currently has special cases of drift for certain fields and behavior fields. These fields do not rely on the hashing mechanism, so hash versioning will not make any changes to these fields.
+
+## Proposed changes
+
+#### Add a CRD hash in the existing drift hash 
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdefj-ghijkl
+    ...
+---
+apiVersion: karpenter.sh/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: default
+  annotations: 
+    karpenter.k8s.aws/ec2nodeclass-hash: mnopqr-stuvwx
+    ...
+---
+apiVersion: karpenter.sh/v1beta1
+kind: Nodeclaim 
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdef-ghijkl
+    karpenter.k8s.aws/ec2nodeclass-hash: mnopqr-stuvwx
+    ...
+```
+
+As the Karpenter CRDs change, Karpenter’s underlying hashing function may also change. To handle the upgrade path, Karpenter will prepend a hash of the CRD to the customer resource spec hash. Karpenter will assert that this struct version is equivalent before checking the customer resource spec hash. 
+
+As a result, two annotations will be used analyze and orchestrate drift: 
+
+* nodepool-hash: (nodepool-crd-version)-(nodepool-hash)
+* ec2nodeclass-hash: (ec2nodeclass-crd-version)-(ec2nodeclass-hash)
+
+### Option 1(Recommended): Fallback to rehashing on the NodeClaim 
+
+Similar to option A, there will be two case for drift:
```

Comment:
What is Option A? I think I missed something

Reply:
Typo from a previous version. Fixed!